File: lib/network/packets/server/s_placeable_list.dart
/// 可放置家具的一個項目。
class PlaceableItem {
  const PlaceableItem({
    required this.propertyId,
    required this.pngid,
    required this.name,
    required this.placement,
    required this.blocking,
    required this.footprintW,
    required this.footprintH,
  });

  final int propertyId;

  /// 圖片編號，對應 object_catalog.json。
  final int pngid;
  final String name;

  /// 放置面：`floor` 需可走格、`wall` 需不可走格、`any` 不限。
  /// 前端據此決定預覽時哪些格子是合法的。
  final String placement;
  final bool blocking;
  final int footprintW;
  final int footprintH;

  bool get isFloor => placement == 'floor';
  bool get isWall => placement == 'wall';

  factory PlaceableItem.fromJson(Map<String, dynamic> j) => PlaceableItem(
        propertyId: (j['propertyId'] as num?)?.toInt() ?? 0,
        pngid: (j['pngid'] as num?)?.toInt() ?? 0,
        name: j['name'] as String? ?? '',
        placement: j['placement'] as String? ?? 'floor',
        blocking: j['blocking'] as bool? ?? false,
        footprintW: (j['footprintW'] as num?)?.toInt() ?? 1,
        footprintH: (j['footprintH'] as num?)?.toInt() ?? 1,
      );
}

/// 可放置家具清單（洞府布置面板用）。
///
/// 只含 `property.placeable = 1` 的項目 —— 樹木、礦石那類世界物件
/// 不會出現在玩家的布置清單裡。
class SPlaceableList {
  const SPlaceableList({required this.items});

  final List<PlaceableItem> items;

  factory SPlaceableList.fromData(Map<String, dynamic> data) {
    final raw = data['items'] as List<dynamic>? ?? const [];
    return SPlaceableList(
      items: raw
          .map((e) => PlaceableItem.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: src/main/java/com/xin/server/packet/server/S_PlaceableList.java
package com.xin.server.packet.server;

import java.util.List;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.xin.server.datatables.PropertyTable;
import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;
import com.xin.server.template.PropertyTemplate;

/**
 * 可放置家具清單（對應 opcode {@code S_PLACEABLE_LIST}）。
 * <p>
 * 洞府布置面板用來顯示玩家能擺哪些東西。只列出 {@code property.placeable = 1} 的項目
 * —— 樹木、礦石那類世界物件不會出現在玩家的布置清單裡。
 * <p>
 * {@code placement} 決定該物件能放在什麼格子上（見 {@code PlacementSurface}），
 * 前端據此決定預覽時哪些格子是合法的。
 *
 * <pre>
 * {
 *   "op": "S_PLACEABLE_LIST",
 *   "data": {
 *     "items": [
 *       { "propertyId": 1200, "pngid": 1200, "name": "木桌一",
 *         "placement": "floor", "blocking": true,
 *         "footprintW": 1, "footprintH": 1 }
 *     ]
 *   }
 * }
 * </pre>
 */
public class S_PlaceableList extends ServerBasePacket {

    private S_PlaceableList() {
        super(ServerOpcodes.S_PLACEABLE_LIST);

        List<PropertyTemplate> items = PropertyTable.get().getPlaceable();
        ArrayNode arr = newArray();
        for (PropertyTemplate t : items) {
            ObjectNode node = newObject();
            node.put("propertyId", t._id);
            node.put("pngid",      t._pngId);
            node.put("name",       t._viewNote);
            node.put("placement",  t._placement != null ? t._placement : "floor");
            node.put("blocking",   t._blocking);
            node.put("footprintW", Math.max(1, t._footprintW));
            node.put("footprintH", Math.max(1, t._footprintH));
            arr.add(node);
        }
        putArray("items", arr);
    }

    public static S_PlaceableList of() {
        return new S_PlaceableList();
    }
}
